src/storage/compression/bitpacking.cpp

```diff
@@ -949,16 +973,70 @@ void BitpackingSkip(ColumnSegment &segment, ColumnScanState &state, idx_t skip_c
 	scan_state.Skip(segment, skip_count);
 }
 
+//===--------------------------------------------------------------------===//
+// Serialization & Cleanup
+//===--------------------------------------------------------------------===//
+SerializedBitpackingSegmentState::SerializedBitpackingSegmentState() {
+}
+
+SerializedBitpackingSegmentState::SerializedBitpackingSegmentState(map<BitpackingMode, idx_t> counts_p) {
+	counts = std::move(counts_p);
+}
+
+void SerializedBitpackingSegmentState::Serialize(Serializer &serializer) const {
+	serializer.WriteProperty(1, "counts", counts);
+}
+
+template <class T>
+unique_ptr<CompressedSegmentState> BitpackingInitSegment(ColumnSegment &segment, block_id_t block_id,
+                                                         optional_ptr<ColumnSegmentState> segment_state) {
+	if (block_id == INVALID_BLOCK) {
+		// Return an empty bitpacking segment state, we'll populate this during compression
+		return make_uniq<BitpackingSegmentState>();
+	}
+	auto result = make_uniq<BitpackingSegmentState>();
+	if (segment_state) {
+		auto &serialized_state = segment_state->Cast<SerializedBitpackingSegmentState>();
+		result->counts = std::move(serialized_state.counts);
+	}
+	return std::move(result);
+}
+
+template <class T>
+unique_ptr<ColumnSegmentState> SerializeState(ColumnSegment &segment) {
+	auto &state = segment.GetSegmentState()->Cast<BitpackingSegmentState>();
+	return make_uniq<SerializedBitpackingSegmentState>(state.counts);
+}
+
+template <class T>
+unique_ptr<ColumnSegmentState> DeserializeState(Deserializer &deserializer) {
+	auto result = make_uniq<SerializedBitpackingSegmentState>();
+	deserializer.ReadProperty(1, "counts", result->counts);
+	return std::move(result);
+}
+
+void CleanupState(ColumnSegment &segment) {
```

Comment:
I don't think we need this, cleanup state is necessary for state that holds lifetimes outside of the scope of the segment, such as additional blocks that it has created - which is what the `UncompressedStringSegmentState` uses it for.

UncompressedString can create additional blocks to hold "overflow" strings, which are bigger than a certain threshold (I think it's 4096 by default)
When the column segment that these additional blocks are a part of gets dropped, we need to inform the block manager that these additional blocks can now also be safely reused.

Reply:
Of course, should have realized. Will remove.

Reply:
Just out of curiosity - we now write extra information, could it happen that this requires an extra block (in the case where everything was filled completely with data and I want to write out the bitmode histogram) or does this extra serialization data get written to a separate place in storage?